### PR TITLE
percentile_circles endpoint added to api

### LINKED ISSUE #23: Return `stopcorr` percentile circles from API endpoint
Related to stop correspondence analysis:

Return percentile circle geometries from a GET endpoint such as `/percentile_circles/<stop_id>`.
This will make visual reporting a lot easier on the client side.
Radius values are already contained in stop_medians as an array, but it's pretty tedious to build the circle geometries using the center point and the radius values on the client side.

The underlying db view is already defined [here](https://github.com/HSLdevcom/hfp-analytics/blob/486db28cb6d27392ad60fd9cb955fe08ba759dc3/db/sql/300_result_access.sql#L45), just need to move it to the `stopcorr` schema and correct SQL file, after related `stopcorr` table definitions.
And then develop the GET endpoint on Python side.

I'd say we could add the `stop_id` as a required parameter the same way we do for HFP observation point clouds, since there is no need to visualize cirles of all stops at once.

### PR DESCRIPTION
**Pull request by Divinare**
Closes #23

### CODE REVIEW

File: db/sql/400_api.sql [hunk 1]
```diff
@@ -21,8 +20,9 @@ CREATE VIEW api.view_jore_stop_4326 AS (
   SELECT cast(ST_AsGeoJSON(sc.*) AS json)
   FROM selected_cols AS sc
 );
+COMMENT ON VIEW api.view_jore_stop_4326 IS
+'Returns jore_stops as GeoJSON features';
```
Comment (datarttu) [inline db/sql/400_api.sql:24]: Suggested: `'Returns all jore_stops as GeoJSON features';`

File: db/sql/400_api.sql [hunk 2]
```diff
@@ -51,8 +51,9 @@ CREATE VIEW api.view_stop_median_4326 AS (
   SELECT cast(ST_AsGeoJSON(sc.*) AS json)
   FROM selected_cols AS sc
 );
+COMMENT ON VIEW api.view_stop_median_4326 IS
+'Returns stop_medians as GeoJSON features';
```
Comment (datarttu) [inline db/sql/400_api.sql:55]: Suggested: `'Returns all stop_medians as GeoJSON features';`

File: db/sql/400_api.sql [hunk 3]
```diff
@@ -91,4 +93,16 @@ RETURNS setof json as $$
   SELECT cast(ST_AsGeoJSON(sc.*) AS json)
   FROM selected_cols AS sc;
 $$ LANGUAGE SQL VOLATILE;
```
Comment (datarttu) [inline db/sql/400_api.sql:95]: Not touched in this PR, but I think `STABLE` can be used in this function for potentially better query optimization than with stricter `VOLATILE`. `STABLE` allows for database lookups (which `IMMUTABLE` does not) that we do here, this is just a "read-only" function whose result depends on the argument values and the database state. `VOLATILE` would be needed if we modified the database with the function and/or used values from functions like `random()`.

You can try changing this within this PR or just add a TODO.

File: db/sql/400_api.sql [hunk 4]
```diff
@@ -91,4 +93,16 @@ RETURNS setof json as $$
...
+COMMENT ON FUNCTION api.get_observations_with_null_stop_id_4326 IS
+'Returns observations with null stop_id as GeoJSON features';
```
Comment (datarttu) [inline db/sql/400_api.sql:97]: Suggested: `'Returns observations with no stop_id value, but located within search_distance_m from the given stop_id, as GeoJSON features';`

File: db/sql/400_api.sql [hunk 5]
```diff
@@ -91,4 +93,16 @@ RETURNS setof json as $$
...
 
+CREATE OR REPLACE FUNCTION api.get_percentile_circles_with_stop_id(stop_id int)
+RETURNS setof json as $$
+  WITH selected_cols AS (
+    SELECT * FROM stopcorr.view_percentile_circles pc WHERE pc.stop_id = $1
+  )
+  SELECT cast(ST_AsGeoJSON(sc.*) AS json)
+  FROM selected_cols AS sc;
+$$ LANGUAGE SQL VOLATILE;
```
Comment (datarttu) [inline db/sql/400_api.sql:106]: Same here as above, read-only function that can be `STABLE` instead of `VOLATILE`.

File: db/sql/400_api.sql [hunk 6]
```diff
@@ -91,4 +93,16 @@ RETURNS setof json as $$
...
+COMMENT ON FUNCTION api.get_percentile_circles_with_stop_id IS
+'Returns percentile circles with stop_id as GeoJSON features';
```
Comment (datarttu) [inline db/sql/400_api.sql:108]: Suggested: `'Returns percentile circles around given stop_id as GeoJSON features';`

File: python/api/main.py [hunk 7]
```diff
@@ -167,9 +167,26 @@ async def get_hfp_points(stop_id: str):
...
                 observation = observation_tuple[0]
                 observation_geojson_features.append(observation)
 
     return get_feature_collection(observation_geojson_features)
+
+
+@app.get("/percentile_circles/{stop_id}")
+async def get_percentile_circles(stop_id: str):
+    """
+    Returns a GeoJSON FeatureCollection of percentile circles with given stop_id.
```
Comment (datarttu) [inline python/api/main.py:180]: Suggested: `Returns a GeoJSON FeatureCollection of percentile circles around the given stop by stop_id.`

File: python/api/main.py [hunk 8]
```diff
@@ -167,9 +167,26 @@ async def get_hfp_points(stop_id: str):
...
+    """
+    with psycopg.connect(**get_conn_params()) as conn:
+        with conn.cursor() as cur:
+            cur.execute("SELECT api.get_percentile_circles_with_stop_id(%(stop_id)s)", {'stop_id': stop_id })
+            percentile_circles = cur.fetchall()
+
+            percentile_circle_geojson_features = []
```
Comment (datarttu) [inline python/api/main.py:187]: Seems like we have a repeated pattern here, common for all these endpoints returning a FeatureCollection. At the same time, the current `get_feature_collection()` func does pretty little compared to this pattern. Maybe we could wrap the whole thing to a function in `common.utils`, e.g. like this?

```python
def tuples_to_feature_collection(geom_tuples: list[tuple]) -> dict
    """Transforms GeoJSON tuples returned by psycopg into a FeatureCollection dict for REST API."""
    # Effectively the same as the list initialization + for loop phase:
    features = [tp[0] for tp in geom_tuples]
    # Then what the current get_feature_collection() does (making that function obsolete now):
    return {
        "type": "FeatureCollection",
        "features": features
    }
```

which would be used here like

```python
return tuples_to_feature_collection(percentile_circles)
```

**Review by datarttu — CHANGES_REQUESTED** [review]
The core functionality looks really good, just commented some metadata nuances, Postgres function optimization, and possibly wrapping the FeatureCollection building process into a function in Python side.

**Comment by Divinare** [comment]
Good points, fixed everything as suggested.

**Comment by datarttu** [comment]
Looks good, think this is ready for merge!

**Pull request merged**